pennylane/plugins/expt_tensornet.py

```diff
@@ -0,0 +1,197 @@
+        print("\n")
+
+    def expval(self, observable, wires, par):
+        if not isinstance(observable, list):
+            observable = [observable]
+            wires = [wires]
+            par = [par]
+        matrices = []
+        for o, p, w in zip(observable, par, wires):
+            A = self._get_operator_matrix(o, p)
+            num_mult_idxs = len(w)
+            matrices.append(np.reshape(A, [2] * num_mult_idxs * 2))
+        nodes = [self._add_node(A, w, name=o) for A, w, o in zip(matrices, wires, observable)]
+        return self.ev(nodes, wires)
+
+    def _get_operator_matrix(self, operation, par):
+        """Get the operator matrix for a given operation or observable.
+
+        Args:
+          operation    (str): name of the operation/observable
+          par (tuple[float]): parameter values
+        Returns:
+          array: matrix representation.
+        """
+        A = {**self._operation_map, **self._observable_map}[operation]
+        if not callable(A):
+            return A
+        return A(*par)
+
+    def ev(self, obs_nodes, wires):
+        r"""Expectation value of observables on specified wires.
+
+         Args:
+            obs_nodes (tn.Node): the observables as tensornetwork Nodes
+            wires (Sequence[int]): measured subsystems for each observable
+         Returns:
+            float: expectation value :math:`\expect{A} = \bra{\psi}A\ket{\psi}`
+        """
+        # first need to connect together nodes representing measurement
+        all_wires = tuple(w for w in range(self.num_wires))
+        ket = self._add_node(self._state, wires=all_wires, name="Ket")
+        bra = self._add_node(tn.conj(ket), wires=all_wires, name="Bra")
+        meas_wires = []
+        for obs_node, obs_wires in zip(obs_nodes, wires):
+            meas_wires.extend(obs_wires)
+            for idx, w in enumerate(obs_wires):
+                left_mult_index = idx
+                right_mult_index = len(obs_wires) + idx
+                self._add_edge(obs_node, right_mult_index, ket, w)  # A|psi>
+                self._add_edge(bra, w, obs_node, left_mult_index)  # <psi|A
+        for w in set(all_wires) - set(meas_wires):
+            self._add_edge(bra, w, ket, w)  # |psi[w]|**2
+
+        # contractions
+        contracted_ket = ket
+        for obs_node in obs_nodes:
+            contracted_ket = tn.contract_between(obs_node, contracted_ket)
+        expval = tn.contract_between(bra, contracted_ket).tensor
+        if np.abs(expval.imag) > tolerance:
+            warnings.warn('Nonvanishing imaginary part {} in expectation value.'.format(expval.imag), RuntimeWarning)
+        return expval.real
+
+    def reset(self):
+        """Reset the device"""
+        self.__init__(self.num_wires)
+
+    @property
+    def operations(self):
+        return set(self._operation_map.keys())
+
+    @property
+    def observables(self):
+        return set(self._observable_map.keys())
```

Comment:
I can't quite remember off the top of my head, but the metric_tensor functionality requires a probability method (I think).

Reply:
Yep, a feature for the future.